### DISCUSSION TITLE
Can someone explain what this release strategy is?

### DISCUSSION THREAD

**Discussion by hobbsh** [apollographql/router#1784] [General]
**Describe the solution you'd like**
I am thoroughly confused by ya'lls release strategy. 

- A breaking helm chart change was released today
- I'm not sure if we should be using the alpha releases, given they're tagged with `alpha`
- Months ago, router was claimed to be GA, but back to the `alpha` tagging, did things regress?

We are pinned to v0.16.0 at the moment which is not compatible with the most recent helm chart version (the `supergraph` configuration key). Please explain what's going on here and how we should follow updates. We are unfortunately not able to use the OCI Helm Repo and version properly so I guess we'll have to resort to hosting the chart ourselves for the time being?

**Comment by hobbsh** [comment]
It's incredibly frustrating that you've opted to only support OCI helm registries which are not well adopted/supported in the ecosystem. I would like to propose that a standard Helm Registry is created and utilized instead.

**Comment by abernix** [comment]
I'll convert this to a discussion as I think it's more of that than an issue based on what you've offered.  There's a bit to unpack here, but at a high-level, I'll say that we aim to have more stable versions of the Router published once we reach 1.0.0, but I also don't believe that's unintuitive considering we're using [Semantic Versioning 2.0](https://semver.org/) on the project, [as indicated at the top of our `CHANGELOG.md`](https://github.com/apollographql/router/blob/main/CHANGELOG.md).

More details in the discussion.  Thanks!

**Comment by abernix** [comment]
At a more detailed level to my above response, I think it will be valuable for posterity to respond to your individual points in case someone is curious about this in the future, so I'll take the time to respond within:

> @hobbsh wrote:
> * A breaking helm chart change was released today

The helm chart released today is `1.0.0-rc.0` (and corresponds to an actual binary / Git tag / release, too).  It is a follow-up release to `1.0.0-alpha.3`.  Also, the version numbers of the Helm chart and the underlying binary/image are in lock-step now in an attempt to reduce cognitive overhead.

Both `rc` and `alpha` are **pre-release** version identifiers per the [semver 2.0 specification](https://semver.org/#spec-item-9):

> _A pre-release version MAY be denoted by appending a hyphen and a series of dot separated identifiers immediately following the patch version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty. Numeric identifiers MUST NOT include leading zeroes. Pre-release versions have a lower precedence than the associated normal version._

It goes on to indicate:

> _A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version. Examples: 1.0.0-alpha, 1.0.0-alpha.1, 1.0.0-0.3.7, 1.0.0-x.7.z.92, 1.0.0-x-y-z.–._

We note our commitment to using semantic versioning to convey the stability of our releases at the top of [our `CHANGELOG.md`](https://github.com/apollographql/router/blob/main/CHANGELOG.md) and it would seem Helm [also claims to abide by Sem Ver 2.0 specification](https://helm.sh/docs/chart_best_practices/conventions/#version-numbers), too.

> _Wherever possible, Helm uses [SemVer 2](https://semver.org/) to represent version numbers. (Note that Docker image tags do not necessarily follow SemVer, and are thus considered an unfortunate exception to the rule.)_

Going back to the above statement, this wording indicates a breaking change can _simply exist_ in pre-release versions.  [Our `CHANGELOG.md`](https://github.com/apollographql/router/blob/main/CHANGELOG.md) and [corresponding GitHub Release](https://github.com/apollographql/router/releases) tags tend to be a useful guide to helping you traverse those changes.  Therefore, your statement that a breaking change was released today in a pre-release version is absolutely true but would seem fully normal, acceptable and defendable for pre-release software versions.

If Helm tooling mis-led you into thinking that there would be no breaking change on a pre-release version or didn't pin the version to a range that you specified, that seems like it could be a failure in Helm's tooling or versioning strategy or your usage of that tooling?  I suspect if you remained on version 0.9.0 things and didn't bump to `1.0.0-alpha.*` or `1.0.0-rc.0` that things would have continued to work as they were for you.

--- 

> @hobbsh wrote:
> * I'm not sure if we should be using the alpha releases, given they're tagged with `alpha`

I believe the terminology "alpha" is common convention in software along with "beta" and "prerelease".  You shouldn't feel compelled to use alpha releases if you don't want to.  You should evaluate your own comfort level with adopting versions of software.  I think thankfully, you can use semantic versioning as a guide to help interpreting that comfort.  Known issues are opened by the community on this repository to highlight specific challenges and up-voted accordingly when they are shared by others, so you can evaluate those when making a decision.  We tend to prioritize the most painful things and course-correct accordingly.

---
> @hobbs wrote:
> * Months ago, router was claimed to be GA, but back to the alpha tagging, did things regress?

Router 0.9.0 was indeed GA — which means general availability.  That "GA" designator doesn't negate the semantic versioning version numbers which we have applied to our releases.  Referring back to the [semantic versioning specification](https://semver.org/#spec-item-4) again:

>  _Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable._

The `alpha` pre-releases are part of our push to get to 1.0 and the release of `1.0.0-rc.0` is another step in that direction.  For the above reasons, they are subject to breaking changes.  We have not made any breaking changes to 0.9.x versions since it was releases even if semantic versioning would fully support us doing that.

---
> @hobbs wrote:
> Please explain what's going on here and how we should follow updates.

We're not doing anything esoteric or ad-hoc here: Our updates and changes are on our [releases](https://github.com/apollographql/router/releases) page and duplicately outlined [on our `CHANGELOG.md`](https://github.com/apollographql/router/blob/main/CHANGELOG.md).  I believe these are common patterns.

I'll flag that this comment seems a bit demanding to me.  When we're openly abiding by common conventions like semantic versioning and CHANGELOGs in expected locations, I'll claim the burden of highlighting discrepancies to common convention is on you.  I suggest you look holistically at the deltas you're observing compared to other software projects rather than analyzing us and questioning us in isolation.  At a meta level, since I do believe we're obeying common conventions in software releasing, I am going to ask you to not come here requesting that we explain those things to you.  If you feel differently, I'm curious why.

---
> @hobbs wrote:
> We are unfortunately not able to use the OCI Helm Repo and version properly so I guess we'll have to resort to hosting the chart ourselves for the time being?

You're welcome to host the chart yourselves, but this comment seems off-topic and unrelated to this issue and more along the lines of another issue you've previously opened:

* #1643 

The [outcome of that](https://github.com/apollographql/router/issues/1643#issuecomment-1231313185) seemed to indicate that `helm repo add` did work properly with an OCI Helm Repo and that perhaps `flux` was unable to work with them.  We didn't hear back from you on that issue, but curious if you found that to be the case or if there was something more to it.  Helm offered OCI images as an option and it's an appealing choice to us for reasons you may or may not be interested in.  Either way, seems like a separate discussion that might be worth having!

---

I hope the above is overall helpful information in understanding the releasing strategy we have employed.  If you feel my claims are off, I'm interested in understanding better.

As a last point that is not specifically about releasing strategy, this is the second time now that I've observed you using the language that I find to be a bit insensitive on this repository.  There are people behind this GitHub repository working on this software and even if we weren't shipping software in the prescribed way I described above, coming here and opening issues with titles like ["Using the OCI Helm Repository is impossible"](https://github.com/apollographql/router/issues/1643) is, I think, lacking some empathy in the words that were chosen.

In that other case my feeling is that the title seems editorialized, it's phrased as a frustration, and the root cause might be something else entirely.  (Personally, I can't subscribe to it being fundamentally impossible, [because Helm built it and seems to support OCI-based charts](https://helm.sh/blog/storing-charts-in-oci/) with `helm`.   Perhaps you meant to open a discussion with the title "need help using Helm charts with Flux"?  Or maybe that is an issue for another project entirely?).

This new topic then uses phrasing that I think pushes us to explain the way we're shipping releases as if they're particularly different without even claiming how they're different or offering concrete or actionable suggestions for improvement about the strategy for us to reason about.

As a reminder, we have a [code of conduct](https://github.com/apollographql/.github/blob/a82b8263fcf4c3c66b7e350939c6ca3a32355566/CODE_OF_CONDUCT.md) that is in force on this entire organization and would prefer a bias toward positivity.

**Reply by hobbsh** [reply]
@abernix I appreciate the details but I think what needs to happen is not an in-depth explanation of topics I already understand, but a roadmap or more communication to your end users. I would appreciate less condescension and more documentation and communication with your end-users. For example, where are any of these transitions and nuances apparent to someone looking at your repository for the first time?

I don't need to go back and forth with you on the things you've brought up. In some case you're right, in some you arent. In all cases, I've felt dismissed. The issues you've brought up in the past were closed without any resolution in my opinion. What do you want me to do when all it seems like you care about is controlling issue count?

You're entirely correct that the title is phrase by frustration. Why don't you try to find out why that might be? I have been frustrated from the start working in Apollo's ecosystem (see the entire arm64 binary debacle, which by the way is still in progress) and I have been repeatedly dismissed on that issue and several others. I understand you're supporting an open source project but you're also a private company and ultimately want people to pay for your projects. I don't understand the dismissal and frankly rude behavior I've observed communicating with Apollo's projects. Maybe there is a thread to pull there? Why don't we start there?

**Reply by hobbsh** [reply]
I would also like to apologize for anything I have said that has been lacking empathy or has seemed insensitive. That was not my intention at all. I know we're all busy people and we're all trying our best. Hopefully we can find an understanding and move forward.
